File: packages/os-domain/src/access-domain/node.ts
import type { AccessRealm } from "./realm.js";
import { refuseAccessDomain } from "./realm.js";
import type { DomainLifetime } from "./temporal.js";
import { PERMANENT, assertLifetimeWellFormed } from "./temporal.js";

/**
 * The longest a domain slug may be. Matches the DNS label limit, because a
 * domain path is the kind of thing that ends up in a URL, a filename, and an
 * operator's terminal.
 */
export const MAX_DOMAIN_SLUG_CHARS = 63;

/**
 * The longest a human-facing domain name may be. Long enough for a real team
 * name, short enough that a path stays readable on a phone.
 */
export const MAX_DOMAIN_DISPLAY_NAME_CHARS = 128;

/** Whether a domain accepts control from above. */
export type InheritanceMode = "inherit" | "isolated";

/**
 * A domain's link to a vault, carrying the project its envelopes are sealed
 * against.
 *
 * The `projectId` is not a convenience copy: `crates/human-vault` binds it into
 * every envelope's AEAD associated data (ADR 0038), so it is part of the
 * ciphertext's identity. `bindVault` is the only constructor and takes the
 * project from the realm, so a binding cannot disagree with where it lives.
 */
export interface VaultBinding {
  readonly vaultId: string;
  readonly projectId: string;
}

export function bindVault(realm: AccessRealm, vaultId: string): VaultBinding {
  return { vaultId, projectId: realm.projectId };
}

/** Refuse a vault sealed against another project. */
export function assertBindingMatchesRealm(
  binding: VaultBinding,
  realm: AccessRealm,
): void {
  if (binding.projectId === realm.projectId) return;
  throw refuseAccessDomain(
    "vault_binding",
    `Vault ${binding.vaultId} is sealed against ${binding.projectId}, realm is ${realm.projectId}`,
    { vaultId: binding.vaultId, sealedAgainst: binding.projectId },
  );
}

/** A node in a realm's access-domain forest. */
export interface AccessDomain {
  readonly id: string;
  readonly realm: AccessRealm;
  /** Absent for a root. A forest has several. */
  readonly parentId?: string;
  readonly slug: string;
  readonly displayName: string;
  readonly inheritance: InheritanceMode;
  readonly lifetime: DomainLifetime;
  readonly vaultBinding?: VaultBinding;
  readonly createdAt: Date;
}

/** What a caller supplies; everything else has a default. */
export interface NewAccessDomain {
  readonly id: string;
  readonly realm: AccessRealm;
  readonly parentId?: string;
  readonly slug: string;
  readonly displayName: string;
  readonly createdAt: Date;
  readonly inheritance?: InheritanceMode;
  readonly lifetime?: DomainLifetime;
  readonly vaultBinding?: VaultBinding;
}

/** An access domain under construction, before it is handed out as readonly. */
type MutableAccessDomain = {
  -readonly [K in keyof AccessDomain]: AccessDomain[K];
};

/** The same, for the input record a rebuild assembles. */
type MutableNewAccessDomain = {
  -readonly [K in keyof NewAccessDomain]: NewAccessDomain[K];
};

/**
 * Build a domain, checking everything a bare node can check.
 *
 * A node knows nothing about the forest, so parent existence, depth, sibling
 * uniqueness and the lifetime-under-parent rule are all checked at insert.
 */
export function makeAccessDomain(input: NewAccessDomain): AccessDomain {
  assertSlug(input.slug);
  assertDisplayName(input.displayName);
  const lifetime = input.lifetime ?? PERMANENT;
  assertLifetimeWellFormed(lifetime, input.createdAt);
  if (input.vaultBinding !== undefined) {
    assertBindingMatchesRealm(input.vaultBinding, input.realm);
  }
  // Optional fields are assigned only when present: this repo runs
  // `exactOptionalPropertyTypes`, so `parentId: undefined` is not the same
  // type as an absent parent, and a root is an absent parent.
  const domain: MutableAccessDomain = {
    id: input.id,
    realm: input.realm,
    slug: input.slug,
    displayName: input.displayName,
    inheritance: input.inheritance ?? "inherit",
    lifetime,
    createdAt: input.createdAt,
  };
  if (input.parentId !== undefined) domain.parentId = input.parentId;
  if (input.vaultBinding !== undefined) {
    domain.vaultBinding = input.vaultBinding;
  }
  return domain;
}

export function isRoot(domain: AccessDomain): boolean {
  return domain.parentId === undefined;
}

/**
 * The same domain under a different parent, or under none.
 *
 * Rebuilt field by field rather than spread-and-deleted: with
 * `exactOptionalPropertyTypes`, a root is an *absent* `parentId`, not one set
 * to `undefined`, and re-running `makeAccessDomain` re-checks the node on the
 * way out. Whether the move is legal is the forest's question.
 */
export function reparented(
  domain: AccessDomain,
  parentId?: string,
): AccessDomain {
  const input: MutableNewAccessDomain = {
    id: domain.id,
    realm: domain.realm,
    slug: domain.slug,
    displayName: domain.displayName,
    createdAt: domain.createdAt,
    inheritance: domain.inheritance,
    lifetime: domain.lifetime,
  };
  if (parentId !== undefined) input.parentId = parentId;
  if (domain.vaultBinding !== undefined) {
    input.vaultBinding = domain.vaultBinding;
  }
  return makeAccessDomain(input);
}

/**
 * Slug rules: a lowercase DNS-ish label.
 *
 * Narrow on purpose. A slug appears in a path, so anything that could be read
 * as path structure (`/`, `.`, `..`) or that differs only by case is refused
 * rather than normalized — two domains whose paths differ only in case are two
 * domains an operator cannot tell apart.
 */
export function assertSlug(slug: string): void {
  if (slug.length === 0) {
    throw refuseAccessDomain("invalid", "Domain slug is empty");
  }
  if ([...slug].length > MAX_DOMAIN_SLUG_CHARS) {
    throw refuseAccessDomain(
      "invalid",
      `Domain slug exceeds ${MAX_DOMAIN_SLUG_CHARS} characters`,
    );
  }
  if (!/^[a-z0-9-]+$/.test(slug)) {
    throw refuseAccessDomain(
      "invalid",
      `Domain slug "${slug}" is not lowercase [a-z0-9-]`,
      {
        slug,
      },
    );
  }
  if (slug.startsWith("-") || slug.endsWith("-")) {
    throw refuseAccessDomain(
      "invalid",
      `Domain slug "${slug}" starts or ends with a hyphen`,
      { slug },
    );
  }
}

/**
 * Display-name rules: present, bounded, single-line. A name with a newline in
 * it can misrepresent a listing.
 */
export function assertDisplayName(displayName: string): void {
  if (displayName.trim().length === 0) {
    throw refuseAccessDomain("invalid", "Domain display name is blank");
  }
  if ([...displayName].length > MAX_DOMAIN_DISPLAY_NAME_CHARS) {
    throw refuseAccessDomain(
      "invalid",
      `Domain display name exceeds ${MAX_DOMAIN_DISPLAY_NAME_CHARS} characters`,
    );
  }
  if (/[\p{Cc}\p{Cf}]/u.test(displayName)) {
    throw refuseAccessDomain(
      "invalid",
      "Domain display name carries a control character",
    );
  }
}

File: crates/domain/src/access_domain/node.rs
//! One access domain: its identity, its slug, and the fields a forest's
//! invariants depend on.
//!
//! The fields are private and reachable only through accessors. That is
//! deliberate: `realm`, `parent_id`, `slug` and `lifetime` are exactly the
//! values [`super::forest::AccessDomainForest`] validated on the way in, so
//! leaving them writable would let a caller edit a forest into a shape no
//! mutation would have accepted — a domain reparented across realms by
//! assignment, for instance. Changing any of them goes through
//! [`super::mutate`], which re-checks.

use super::bridge::VaultBinding;
use super::control::InheritanceMode;
use super::realm::Realm;
use super::temporal::DomainLifetime;
use crate::{AccessDomainId, DomainError};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// The longest a domain slug may be. Matches the DNS label limit, because a
/// domain path is the kind of thing that ends up in a URL, a filename, and an
/// operator's terminal.
pub const MAX_DOMAIN_SLUG_CHARS: usize = 63;

/// The longest a human-facing domain name may be. Long enough for a real team
/// name, short enough that a path stays readable on a phone.
pub const MAX_DOMAIN_DISPLAY_NAME_CHARS: usize = 128;

/// A node in a realm's access-domain forest.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct AccessDomain {
    id: AccessDomainId,
    realm: Realm,
    parent_id: Option<AccessDomainId>,
    slug: String,
    display_name: String,
    inheritance: InheritanceMode,
    lifetime: DomainLifetime,
    vault_binding: Option<VaultBinding>,
    created_at: DateTime<Utc>,
}

impl AccessDomain {
    /// A root domain: one of the realm's forest roots, with no parent.
    ///
    /// # Errors
    ///
    /// [`DomainError::AccessDomainInvalid`] when the slug or display name is
    /// malformed.
    pub fn root(
        id: AccessDomainId,
        realm: Realm,
        slug: &str,
        display_name: &str,
        created_at: DateTime<Utc>,
    ) -> Result<Self, DomainError> {
        Self::build(id, realm, None, slug, display_name, created_at)
    }

    /// A child domain under `parent_id`.
    ///
    /// The parent is named but not checked here — a bare node knows nothing
    /// about the forest. Existence, depth, sibling uniqueness, and the lifetime
    /// and vault rules are the forest's, at insert.
    ///
    /// # Errors
    ///
    /// As [`AccessDomain::root`].
    pub fn child(
        id: AccessDomainId,
        realm: Realm,
        parent_id: AccessDomainId,
        slug: &str,
        display_name: &str,
        created_at: DateTime<Utc>,
    ) -> Result<Self, DomainError> {
        Self::build(id, realm, Some(parent_id), slug, display_name, created_at)
    }

    fn build(
        id: AccessDomainId,
        realm: Realm,
        parent_id: Option<AccessDomainId>,
        slug: &str,
        display_name: &str,
        created_at: DateTime<Utc>,
    ) -> Result<Self, DomainError> {
        assert_slug(slug)?;
        assert_display_name(display_name)?;
        Ok(Self {
            id,
            realm,
            parent_id,
            slug: slug.to_string(),
            display_name: display_name.to_string(),
            inheritance: InheritanceMode::Inherit,
            lifetime: DomainLifetime::Permanent,
            vault_binding: None,
            created_at,
        })
    }

    /// Break control inheritance at this domain, or restore it.
    #[must_use]
    pub fn with_inheritance(mut self, inheritance: InheritanceMode) -> Self {
        self.inheritance = inheritance;
        self
    }

    /// Put this domain on a clock.
    ///
    /// # Errors
    ///
    /// [`DomainError::AccessDomainLifetime`] when the deadline is already past
    /// at creation or beyond the cap. Whether it also fits under the parent is
    /// the forest's check, because only the forest knows the parent.
    pub fn with_lifetime(mut self, lifetime: DomainLifetime) -> Result<Self, DomainError> {
        lifetime.assert_well_formed(self.created_at)?;
        self.lifetime = lifetime;
        Ok(self)
    }

    /// Bind a vault to this domain.
    ///
    /// # Errors
    ///
    /// [`DomainError::AccessDomainVaultBinding`] when the binding's project is
    /// not this domain's realm — see [`VaultBinding`] for why that is fatal
    /// rather than correctable.
    pub fn with_vault_binding(mut self, binding: VaultBinding) -> Result<Self, DomainError> {
        binding.assert_matches_realm(&self.realm)?;
        self.vault_binding = Some(binding);
        Ok(self)
    }

    #[must_use]
    pub fn id(&self) -> AccessDomainId {
        self.id
    }

    #[must_use]
    pub fn realm(&self) -> Realm {
        self.realm
    }

    #[must_use]
    pub fn parent_id(&self) -> Option<AccessDomainId> {
        self.parent_id
    }

    #[must_use]
    pub fn is_root(&self) -> bool {
        self.parent_id.is_none()
    }

    #[must_use]
    pub fn slug(&self) -> &str {
        &self.slug
    }

    #[must_use]
    pub fn display_name(&self) -> &str {
        &self.display_name
    }

    #[must_use]
    pub fn inheritance(&self) -> InheritanceMode {
        self.inheritance
    }

    #[must_use]
    pub fn lifetime(&self) -> DomainLifetime {
        self.lifetime
    }

    #[must_use]
    pub fn vault_binding(&self) -> Option<VaultBinding> {
        self.vault_binding
    }

    #[must_use]
    pub fn created_at(&self) -> DateTime<Utc> {
        self.created_at
    }

    /// Whether this domain is still live at the caller's clock reading.
    ///
    /// Answers only for this node. A live domain under an expired ancestor is
    /// unreachable, and `AccessDomainForest::assert_reachable` is the check
    /// that says so.
    ///
    /// # Errors
    ///
    /// [`DomainError::AccessDomainExpired`] once the deadline has passed.
    pub fn assert_active(&self, now: DateTime<Utc>) -> Result<(), DomainError> {
        self.lifetime.assert_active(now)
    }

    pub(super) fn set_parent(&mut self, parent_id: Option<AccessDomainId>) {
        self.parent_id = parent_id;
    }

    pub(super) fn set_slug(&mut self, slug: String) {
        self.slug = slug;
    }

    pub(super) fn set_display_name(&mut self, display_name: String) {
        self.display_name = display_name;
    }

    pub(super) fn set_inheritance(&mut self, inheritance: InheritanceMode) {
        self.inheritance = inheritance;
    }

    pub(super) fn set_lifetime(&mut self, lifetime: DomainLifetime) {
        self.lifetime = lifetime;
    }
}

/// Slug rules: a lowercase DNS-ish label.
///
/// Narrow on purpose. A slug appears in a path, so anything that could be read
/// as path structure (`/`, `.`, `..`) or that differs only by case is refused
/// rather than normalized — two domains whose paths differ only in case are two
/// domains an operator cannot tell apart.
///
/// # Errors
///
/// [`DomainError::AccessDomainInvalid`] for an empty slug, one over
/// [`MAX_DOMAIN_SLUG_CHARS`], one with a character outside `[a-z0-9-]`, or one
/// starting or ending with a hyphen.
pub fn assert_slug(slug: &str) -> Result<(), DomainError> {
    if slug.is_empty() {
        return Err(DomainError::AccessDomainInvalid(
            "domain slug is empty".to_string(),
        ));
    }
    if slug.chars().count() > MAX_DOMAIN_SLUG_CHARS {
        return Err(DomainError::AccessDomainInvalid(format!(
            "domain slug exceeds {MAX_DOMAIN_SLUG_CHARS} characters"
        )));
    }
    if !slug
        .chars()
        .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-')
    {
        return Err(DomainError::AccessDomainInvalid(format!(
            "domain slug {slug:?} is not lowercase [a-z0-9-]"
        )));
    }
    if slug.starts_with('-') || slug.ends_with('-') {
        return Err(DomainError::AccessDomainInvalid(format!(
            "domain slug {slug:?} starts or ends with a hyphen"
        )));
    }
    Ok(())
}

/// Display-name rules: present, bounded, single-line.
///
/// # Errors
///
/// [`DomainError::AccessDomainInvalid`] when blank, over
/// [`MAX_DOMAIN_DISPLAY_NAME_CHARS`], or carrying a control character — a name
/// with a newline in it can misrepresent a listing.
pub fn assert_display_name(display_name: &str) -> Result<(), DomainError> {
    if display_name.trim().is_empty() {
        return Err(DomainError::AccessDomainInvalid(
            "domain display name is blank".to_string(),
        ));
    }
    if display_name.chars().count() > MAX_DOMAIN_DISPLAY_NAME_CHARS {
        return Err(DomainError::AccessDomainInvalid(format!(
            "domain display name exceeds {MAX_DOMAIN_DISPLAY_NAME_CHARS} characters"
        )));
    }
    if display_name.chars().any(char::is_control) {
        return Err(DomainError::AccessDomainInvalid(
            "domain display name carries a control character".to_string(),
        ));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{ProjectId, VaultId};

    fn now() -> DateTime<Utc> {
        DateTime::parse_from_rfc3339("2026-03-01T00:00:00Z")
            .expect("fixture timestamp")
            .with_timezone(&Utc)
    }

    fn realm() -> Realm {
        Realm::standard(None, ProjectId::new())
    }

    #[test]
    fn a_root_has_no_parent_and_inherits_by_default() {
        let domain = AccessDomain::root(
            AccessDomainId::new(),
            realm(),
            "platform",
            "Platform",
            now(),
        )
        .expect("valid root");
        assert!(domain.is_root());
        assert_eq!(domain.parent_id(), None);
        assert_eq!(domain.inheritance(), InheritanceMode::Inherit);
        assert_eq!(domain.lifetime(), DomainLifetime::Permanent);
        assert_eq!(domain.vault_binding(), None);
    }

    #[test]
    fn slugs_refuse_path_structure_and_case() {
        for bad in [
            "",
            "Platform",
            "plat form",
            "plat/form",
            "..",
            "-platform",
            "platform-",
            "platform_prod",
            "plaeform\u{e4}",
        ] {
            assert!(assert_slug(bad).is_err(), "{bad:?} should be refused");
        }
        for good in ["platform", "prod-1", "a", "0"] {
            assert!(assert_slug(good).is_ok(), "{good:?} should be accepted");
        }
    }

    #[test]
    fn slug_length_is_bounded() {
        assert!(assert_slug(&"a".repeat(MAX_DOMAIN_SLUG_CHARS)).is_ok());
        assert!(assert_slug(&"a".repeat(MAX_DOMAIN_SLUG_CHARS + 1)).is_err());
    }

    #[test]
    fn display_names_refuse_blanks_and_control_characters() {
        assert!(assert_display_name("Platform").is_ok());
        assert!(assert_display_name("   ").is_err());
        assert!(assert_display_name("Plat\nform").is_err());
        assert!(assert_display_name(&"a".repeat(MAX_DOMAIN_DISPLAY_NAME_CHARS + 1)).is_err());
    }

    #[test]
    fn a_vault_binding_from_another_realm_is_refused_at_the_node() {
        let realm = realm();
        let elsewhere = Realm::standard(None, ProjectId::new());
        let domain =
            AccessDomain::root(AccessDomainId::new(), realm, "platform", "Platform", now())
                .expect("valid root");
        let foreign = VaultBinding::bind(&elsewhere, VaultId::new());
        assert!(domain.clone().with_vault_binding(foreign).is_err());
        let own = VaultBinding::bind(&realm, VaultId::new());
        assert!(domain.with_vault_binding(own).is_ok());
    }

    #[test]
    fn a_spent_lifetime_is_refused_at_construction() {
        let domain = AccessDomain::root(AccessDomainId::new(), realm(), "visit", "Visit", now())
            .expect("valid root");
        assert!(domain
            .clone()
            .with_lifetime(DomainLifetime::Temporary { expires_at: now() })
            .is_err());
        assert!(domain
            .with_lifetime(DomainLifetime::Temporary {
                expires_at: now() + chrono::Duration::hours(1),
            })
            .is_ok());
    }
}
